README
= DateHelperJa プラグイン

== DateHelperJa とは

ActionView::Helpers::DateHelper を日本語化する。
DateFormJa をプラグインとしてインストールするだけで、datetime_select,date_selectが日本語表示される。

== インストール

 % ./script/plugin install git://github.com/elm200/date_helper_ja.git

== 使用例

datetime_select("model", "attr") 
# =>"2009年 4月 29日 13時 14分"

オリジナルの振る舞いに戻したいときは、:use_jp_xxx (xxx = year, month, day, hour, minute, second) というオプションを false に設定する。

datetime_select("model", "attr", :use_jp_hour => false, :use_jp_minute => false)
# => "2009年 4月 29日 - 13:14"

:use_era_name というオプションを true にすることで和暦を表示する。

datetime_select("model", "attr", :use_era_name => true)
# => "平成21年 4月 29日 13時14分"

さらに :era_format というオプションで和暦の表示スタイルを指定できる。
:ja_long(デフォルト値), :ja_short, :alphabet が指定可能。

datetime_select("model", "attr", :use_era_name => true, :era_format => :ja_long)
# => "平成21年 4月 29日 13時14分"

datetime_select("model", "attr", :use_era_name => true, :era_format => :ja_short)
# => "平21年 4月 29日 13時14分"

datetime_select("model", "attr", :use_era_name => true, :era_format => :alphabet)
# => "H21年 4月 29日 13時14分"

== アプリケーションテンプレート機能(Rails 2.3より使用可)を使い、サンプルアプリを作成する

 % rails sample -m http://github.com/elm200/date_helper_ja/raw/master/templates/date_helper_ja.rb

== リンク

* Rails 1.x.x に対応した DateHelperJa 旧バージョン: http://d.hatena.ne.jp/elm200/20070430/1177814782
 
== 作者
 
Copyright 2006 - 2009 (c) Eiji Sakai, under MIT License.

Eiji Sakai <eiji.sakai@softculture.com>
 
http://d.hatena.ne.jp/elm200/

CHANGELOG
* 0.2 (2009-07-27)
 Nothing has been changed. Tested that this plugin is compatible with Rails 2.3.2 and 2.3.3.

* 0.2 (2009-02-21)
 Second release.
 Now the plugin can be used with Rails 2.1 and Rails 2.2 (I am not sure with Rails 2.0)
 Era based years(Wareki) can be displayed.
 distance_of_time_in_words method has been changed slightly. 
 
* 0.1 (2008-04-30)
 Initial release
 http://d.hatena.ne.jp/elm200/20070430/1177814782
 

test/date_helper_ja_test.rb
ENV["RAILS_ENV"] = "test"
require File.expand_path(File.dirname(__FILE__) + "/../../../../config/environment")
require 'test_help'

# Show backtraces for deprecated behavior for quicker cleanup.
ActiveSupport::Deprecation.debug = true

ActionController::Base.logger = nil
ActionController::Routing::Routes.reload rescue nil

class DateHelperJaTest < Test::Unit::TestCase
  include ActionView::Helpers::DateHelper
  include ActionView::Helpers::FormHelper

  silence_warnings do
    Post = Struct.new("Post", :id, :written_on, :updated_at)
    Post.class_eval do
      def id
        123
      end
      def id_before_type_cast
        123
      end
    end
  end

  def test_select_year
    expected = %(<select id="date_year" name="date[year]">\n)
    expected << %(<option selected="selected" value="2003">2003</option>\n<option value="2004">2004</option>\n<option value="2005">2005</option>\n)
    expected << "</select>年\n"

    assert_equal expected, select_year(Time.mktime(2003, 8, 16), :start_year => 2003, :end_year => 2005)
    assert_equal expected, select_year(2003, :start_year => 2003, :end_year => 2005)
  end

  def test_select_month
    expected = %(<select id="date_month" name="date[month]">\n)
    expected << %(<option value="1">1</option>\n<option value="2">2</option>\n<option value="3">3</option>\n<option value="4">4</option>\n<option value="5">5</option>\n<option value="6">6</option>\n<option value="7">7</option>\n<option selected="selected" value="8">8</option>\n<option value="9">9</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n)
    expected << "</select>月\n"

    assert_equal expected, select_month(Time.mktime(2003, 8, 16))
    assert_equal expected, select_month(8)
  end

  def test_select_day
    expected = %(<select id="date_day" name="date[day]">\n)
    expected << %(<option value="1">1</option>\n<option value="2">2</option>\n<option value="3">3</option>\n<option value="4">4</option>\n<option value="5">5</option>\n<option value="6">6</option>\n<option value="7">7</option>\n<option value="8">8</option>\n<option value="9">9</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option value="15">15</option>\n<option selected="selected" value="16">16</option>\n<option value="17">17</option>\n<option value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n<option value="24">24</option>\n<option value="25">25</option>\n<option value="26">26</option>\n<option value="27">27</option>\n<option value="28">28</option>\n<option value="29">29</option>\n<option value="30">30</option>\n<option value="31">31</option>\n)
    expected << "</select>日\n"

    assert_equal expected, select_day(Time.mktime(2003, 8, 16))
    assert_equal expected, select_day(16)
  end

  def test_select_hour
    expected = %(<select id="date_hour" name="date[hour]">\n)
    expected << %(<option value="00">00</option>\n<option value="01">01</option>\n<option value="02">02</option>\n<option value="03">03</option>\n<option value="04">04</option>\n<option value="05">05</option>\n<option value="06">06</option>\n<option value="07">07</option>\n<option selected="selected" value="08">08</option>\n<option value="09">09</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option value="15">15</option>\n<option value="16">16</option>\n<option value="17">17</option>\n<option value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n)
    expected << "</select>時\n"

    assert_equal expected, select_hour(Time.mktime(2003, 8, 16, 8, 4, 18))
  end

  def test_select_minute
    expected = %(<select id="date_minute" name="date[minute]">\n)
    expected << %(<option value="00">00</option>\n<option value="01">01</option>\n<option value="02">02</option>\n<option value="03">03</option>\n<option selected="selected" value="04">04</option>\n<option value="05">05</option>\n<option value="06">06</option>\n<option value="07">07</option>\n<option value="08">08</option>\n<option value="09">09</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option value="15">15</option>\n<option value="16">16</option>\n<option value="17">17</option>\n<option value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n<option value="24">24</option>\n<option value="25">25</option>\n<option value="26">26</option>\n<option value="27">27</option>\n<option value="28">28</option>\n<option value="29">29</option>\n<option value="30">30</option>\n<option value="31">31</option>\n<option value="32">32</option>\n<option value="33">33</option>\n<option value="34">34</option>\n<option value="35">35</option>\n<option value="36">36</option>\n<option value="37">37</option>\n<option value="38">38</option>\n<option value="39">39</option>\n<option value="40">40</option>\n<option value="41">41</option>\n<option value="42">42</option>\n<option value="43">43</option>\n<option value="44">44</option>\n<option value="45">45</option>\n<option value="46">46</option>\n<option value="47">47</option>\n<option value="48">48</option>\n<option value="49">49</option>\n<option value="50">50</option>\n<option value="51">51</option>\n<option value="52">52</option>\n<option value="53">53</option>\n<option value="54">54</option>\n<option value="55">55</option>\n<option value="56">56</option>\n<option value="57">57</option>\n<option value="58">58</option>\n<option value="59">59</option>\n)
    expected << "</select>分\n"

    assert_equal expected, select_minute(Time.mktime(2003, 8, 16, 8, 4, 18))
  end

  def test_select_second
    expected = %(<select id="date_second" name="date[second]">\n)
    expected << %(<option value="00">00</option>\n<option value="01">01</option>\n<option value="02">02</option>\n<option value="03">03</option>\n<option value="04">04</option>\n<option value="05">05</option>\n<option value="06">06</option>\n<option value="07">07</option>\n<option value="08">08</option>\n<option value="09">09</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option value="15">15</option>\n<option value="16">16</option>\n<option value="17">17</option>\n<option selected="selected" value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n<option value="24">24</option>\n<option value="25">25</option>\n<option value="26">26</option>\n<option value="27">27</option>\n<option value="28">28</option>\n<option value="29">29</option>\n<option value="30">30</option>\n<option value="31">31</option>\n<option value="32">32</option>\n<option value="33">33</option>\n<option value="34">34</option>\n<option value="35">35</option>\n<option value="36">36</option>\n<option value="37">37</option>\n<option value="38">38</option>\n<option value="39">39</option>\n<option value="40">40</option>\n<option value="41">41</option>\n<option value="42">42</option>\n<option value="43">43</option>\n<option value="44">44</option>\n<option value="45">45</option>\n<option value="46">46</option>\n<option value="47">47</option>\n<option value="48">48</option>\n<option value="49">49</option>\n<option value="50">50</option>\n<option value="51">51</option>\n<option value="52">52</option>\n<option value="53">53</option>\n<option value="54">54</option>\n<option value="55">55</option>\n<option value="56">56</option>\n<option value="57">57</option>\n<option value="58">58</option>\n<option value="59">59</option>\n)
    expected << "</select>秒\n"

    assert_equal expected, select_second(Time.mktime(2003, 8, 16, 8, 4, 18))
  end

  def test_datetime_select
    @post = Post.new
    @post.updated_at = Time.local(2004, 8, 15, 16, 35)

    expected = %{<select id="post_updated_at_1i" name="post[updated_at(1i)]">\n}
    expected << %{<option value="1999">1999</option>\n<option value="2000">2000</option>\n<option value="2001">2001</option>\n<option value="2002">2002</option>\n<option value="2003">2003</option>\n<option selected="selected" value="2004">2004</option>\n<option value="2005">2005</option>\n<option value="2006">2006</option>\n<option value="2007">2007</option>\n<option value="2008">2008</option>\n<option value="2009">2009</option>\n}
    expected << "</select>年\n"

    expected << %{<select id="post_updated_at_2i" name="post[updated_at(2i)]">\n}
    expected << %{<option value="1">1</option>\n<option value="2">2</option>\n<option value="3">3</option>\n<option value="4">4</option>\n<option value="5">5</option>\n<option value="6">6</option>\n<option value="7">7</option>\n<option selected="selected" value="8">8</option>\n<option value="9">9</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n}
    expected << "</select>月\n"

    expected << %{<select id="post_updated_at_3i" name="post[updated_at(3i)]">\n}
    expected << %{<option value="1">1</option>\n<option value="2">2</option>\n<option value="3">3</option>\n<option value="4">4</option>\n<option value="5">5</option>\n<option value="6">6</option>\n<option value="7">7</option>\n<option value="8">8</option>\n<option value="9">9</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option selected="selected" value="15">15</option>\n<option value="16">16</option>\n<option value="17">17</option>\n<option value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n<option value="24">24</option>\n<option value="25">25</option>\n<option value="26">26</option>\n<option value="27">27</option>\n<option value="28">28</option>\n<option value="29">29</option>\n<option value="30">30</option>\n<option value="31">31</option>\n}
    expected << "</select>日\n"

    expected << " "

    expected << %{<select id="post_updated_at_4i" name="post[updated_at(4i)]">\n}
    expected << %{<option value="00">00</option>\n<option value="01">01</option>\n<option value="02">02</option>\n<option value="03">03</option>\n<option value="04">04</option>\n<option value="05">05</option>\n<option value="06">06</option>\n<option value="07">07</option>\n<option value="08">08</option>\n<option value="09">09</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option value="15">15</option>\n<option selected="selected" value="16">16</option>\n<option value="17">17</option>\n<option value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n}
    expected << "</select>時\n"
    expected << %{<select id="post_updated_at_5i" name="post[updated_at(5i)]">\n}
    expected << %{<option value="00">00</option>\n<option value="01">01</option>\n<option value="02">02</option>\n<option value="03">03</option>\n<option value="04">04</option>\n<option value="05">05</option>\n<option value="06">06</option>\n<option value="07">07</option>\n<option value="08">08</option>\n<option value="09">09</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option value="15">15</option>\n<option value="16">16</option>\n<option value="17">17</option>\n<option value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n<option value="24">24</option>\n<option value="25">25</option>\n<option value="26">26</option>\n<option value="27">27</option>\n<option value="28">28</option>\n<option value="29">29</option>\n<option value="30">30</option>\n<option value="31">31</option>\n<option value="32">32</option>\n<option value="33">33</option>\n<option value="34">34</option>\n<option selected="selected" value="35">35</option>\n<option value="36">36</option>\n<option value="37">37</option>\n<option value="38">38</option>\n<option value="39">39</option>\n<option value="40">40</option>\n<option value="41">41</option>\n<option value="42">42</option>\n<option value="43">43</option>\n<option value="44">44</option>\n<option value="45">45</option>\n<option value="46">46</option>\n<option value="47">47</option>\n<option value="48">48</option>\n<option value="49">49</option>\n<option value="50">50</option>\n<option value="51">51</option>\n<option value="52">52</option>\n<option value="53">53</option>\n<option value="54">54</option>\n<option value="55">55</option>\n<option value="56">56</option>\n<option value="57">57</option>\n<option value="58">58</option>\n<option value="59">59</option>\n}
    expected << "</select>分\n"

    actual = datetime_select("post", "updated_at")
    #File.open("expected.txt", "w") { |f| f.puts expected }
    #File.open("actual.txt", "w") { |f| f.puts actual }

    assert_equal expected, actual
  end

  def test_datetime_select_with_seconds
    @post = Post.new
    @post.updated_at = Time.local(2004, 8, 15, 16, 35, 5)

    expected = %{<select id="post_updated_at_1i" name="post[updated_at(1i)]">\n}
    expected << %{<option value="1999">1999</option>\n<option value="2000">2000</option>\n<option value="2001">2001</option>\n<option value="2002">2002</option>\n<option value="2003">2003</option>\n<option selected="selected" value="2004">2004</option>\n<option value="2005">2005</option>\n<option value="2006">2006</option>\n<option value="2007">2007</option>\n<option value="2008">2008</option>\n<option value="2009">2009</option>\n}
    expected << "</select>年\n"

    expected << %{<select id="post_updated_at_2i" name="post[updated_at(2i)]">\n}
    expected << %{<option value="1">1</option>\n<option value="2">2</option>\n<option value="3">3</option>\n<option value="4">4</option>\n<option value="5">5</option>\n<option value="6">6</option>\n<option value="7">7</option>\n<option selected="selected" value="8">8</option>\n<option value="9">9</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n}
    expected << "</select>月\n"

    expected << %{<select id="post_updated_at_3i" name="post[updated_at(3i)]">\n}
    expected << %{<option value="1">1</option>\n<option value="2">2</option>\n<option value="3">3</option>\n<option value="4">4</option>\n<option value="5">5</option>\n<option value="6">6</option>\n<option value="7">7</option>\n<option value="8">8</option>\n<option value="9">9</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option selected="selected" value="15">15</option>\n<option value="16">16</option>\n<option value="17">17</option>\n<option value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n<option value="24">24</option>\n<option value="25">25</option>\n<option value="26">26</option>\n<option value="27">27</option>\n<option value="28">28</option>\n<option value="29">29</option>\n<option value="30">30</option>\n<option value="31">31</option>\n}
    expected << "</select>日\n"

    expected << " "

    expected << %{<select id="post_updated_at_4i" name="post[updated_at(4i)]">\n}
    expected << %{<option value="00">00</option>\n<option value="01">01</option>\n<option value="02">02</option>\n<option value="03">03</option>\n<option value="04">04</option>\n<option value="05">05</option>\n<option value="06">06</option>\n<option value="07">07</option>\n<option value="08">08</option>\n<option value="09">09</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option value="15">15</option>\n<option selected="selected" value="16">16</option>\n<option value="17">17</option>\n<option value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n}
    expected << "</select>時\n"
    expected << %{<select id="post_updated_at_5i" name="post[updated_at(5i)]">\n}
    expected << %{<option value="00">00</option>\n<option value="01">01</option>\n<option value="02">02</option>\n<option value="03">03</option>\n<option value="04">04</option>\n<option value="05">05</option>\n<option value="06">06</option>\n<option value="07">07</option>\n<option value="08">08</option>\n<option value="09">09</option>\n<option value="10">10</option>\n<option value="11">11</option>\n<option value="12">12</option>\n<option value="13">13</option>\n<option value="14">14</option>\n<option value="15">15</option>\n<option value="16">16</option>\n<option value="17">17</option>\n<option value="18">18</option>\n<option value="19">19</option>\n<option value="20">20</option>\n<option value="21">21</option>\n<option value="22">22</option>\n<option value="23">23</option>\n<option value="24">24</option>\n<option value="25">25</option>\n<option value="26">26</option>\n<option value="27">27</option>\n<option value="28">28</option>\n<option value="29">29</option>\n<option value="30">30</option>\n<option value="31">31</option>\n<option value="32">32</option>\n<option value="33">33</option>\n<option value="34">34</option>\n<option selected="selected" value="35">35</option>\n<option value="36">36</option>\n<option value="37">37</option>\n<option value="38">38</option>\n<option value="39">39</option>\n<option value="40">40</option>\n<option value="41">41</option>\n<option value="42">42</option>\n<option value="43">43</option>\n<option value="44">44</option>\n<option value="45">45</option>\n<option value="46">46</option>\n<option value="47">47</option>\n<option value="48">48</option>\n<option value="49">49</option>\n<option value="50">50</option>\n<option value="51">51</option>\n<option value="52">52</option>\n<option value="53">53</option>\n<option value="54">54</option>\n<option value="55">55</option>\n<option value="56">56</option>\n<option value="57">57</option>\n<option value="58">58</option>\n<option value="59">59</option>\n}
    expected << "</select>分\n"

    expected << <<-EOS
<select id="post_updated_at_6i" name="post[updated_at(6i)]">
<option value="00">00</option>
<option value="01">01</option>
<option value="02">02</option>
<option value="03">03</option>
<option value="04">04</option>
<option selected="selected" value="05">05</option>
<option value="06">06</option>
<option value="07">07</option>
<option value="08">08</option>
<option value="09">09</option>
<option value="10">10</option>
<option value="11">11</option>
<option value="12">12</option>
<option value="13">13</option>
<option value="14">14</option>
<option value="15">15</option>
<option value="16">16</option>
<option value="17">17</option>
<option value="18">18</option>
<option value="19">19</option>
<option value="20">20</option>
<option value="21">21</option>
<option value="22">22</option>
<option value="23">23</option>
<option value="24">24</option>
<option value="25">25</option>
<option value="26">26</option>
<option value="27">27</option>
<option value="28">28</option>
<option value="29">29</option>
<option value="30">30</option>
<option value="31">31</option>
<option value="32">32</option>
<option value="33">33</option>
<option value="34">34</option>
<option value="35">35</option>
<option value="36">36</option>
<option value="37">37</option>
<option value="38">38</option>
<option value="39">39</option>
<option value="40">40</option>
<option value="41">41</option>
<option value="42">42</option>
<option value="43">43</option>
<option value="44">44</option>
<option value="45">45</option>
<option value="46">46</option>
<option value="47">47</option>
<option value="48">48</option>
<option value="49">49</option>
<option value="50">50</option>
<option value="51">51</option>
<option value="52">52</option>
<option value="53">53</option>
<option value="54">54</option>
<option value="55">55</option>
<option value="56">56</option>
<option value="57">57</option>
<option value="58">58</option>
<option value="59">59</option>
</select>秒
    EOS

    actual = datetime_select("post", "updated_at", :include_seconds => true)

    #File.open("expected.txt", "w") { |f| f.puts expected }
    #File.open("actual.txt", "w") { |f| f.puts actual }

    assert_equal expected, actual
  end

  def test_select_year_with_ja_long_era_format
    expected = %(<select id="date_year" name="date[year]">\n)
    expected << %(<option value=\"1988\">昭和63年</option>\n<option selected=\"selected\" value=\"1989\">昭64年/平1年</option>\n<option value=\"1990\">平成2年</option>\n)
    expected << "</select>\n"

    assert_equal expected, select_year(Time.mktime(1989, 8, 16), :start_year => 1988, :end_year => 1990, :use_era_name => true)
    assert_equal expected, select_year(1989, :start_year => 1988, :end_year => 1990, :use_era_name => true)
  end

  def test_select_year_with_ja_short_era_format
    expected = %(<select id="date_year" name="date[year]">\n)
    expected << %(<option value=\"1988\">昭63</option>\n<option selected=\"selected\" value=\"1989\">昭64/平1</option>\n<option value=\"1990\">平2</option>\n)
    expected << "</select>年\n"

    assert_equal expected, select_year(Time.mktime(1989, 8, 16), :start_year => 1988, :end_year => 1990, :use_era_name => true, :era_format => :ja_short)
    assert_equal expected, select_year(1989, :start_year => 1988, :end_year => 1990, :use_era_name => true, :era_format => :ja_short)
  end

  def test_select_year_with_alpha_era_format
    expected = %(<select id="date_year" name="date[year]">\n)
    expected << %(<option value=\"1988\">S63</option>\n<option selected=\"selected\" value=\"1989\">S64/H1</option>\n<option value=\"1990\">H2</option>\n)
    expected << "</select>年\n"

    assert_equal expected, select_year(Time.mktime(1989, 8, 16), :start_year => 1988, :end_year => 1990, :use_era_name => true, :era_format => :alphabet)
    assert_equal expected, select_year(1989, :start_year => 1988, :end_year => 1990, :use_era_name => true, :era_format => :alphabet)
  end
end


lib/date_helper_ja_2_1_2.rb
# -*- coding: utf-8 -*-
#
# DateHelperJa
#
# Japanizes ActionView::Helpers::DateHelper.
# This is code for Ruby on Rails 2.1.2
#
# Released under the MIT license
# Eiji Sakai <eiji.sakai@softculture.com>
# http://d.hatena.ne.jp/elm200
#
# Hideki Hayashi <hideki.gml@gmail.com>
# http://d.hatena.ne.jp/milk1000cc
#
module ActionView
  module Helpers
    module DateHelper
      #
      # 2つの Time オブジェクト・Date オブジェクトまたは整数(秒数）のだいたいの間隔を報告する。
      # 1分29秒以内の詳細な近似値が必要なら、<tt>include_secondes</tt> を true にすること。
      # 間隔は次の表に基づいて報告される：
      #
      # 0 <-> 29 秒                                                               # => 1分以内
      # 30 秒 <-> 1 分 29 秒                                                      # => 1分
      # 1 分 30 秒 <-> 44 分 29 秒                                                # => [2..44]分
      # 44 分  30 秒 <-> 89 分, 29 秒                                             # => 約1時間
      # 89 分 29 秒 <-> 23 時間 59 分 29 秒                                       # => 約[2..24]時間
      # 23 時間 59 分 29 秒 <-> 47 時間 59 分 29 秒                               # => 1日
      # 47 時間 59 分 29 秒 <-> 29 日 23 時間 59 分 29 秒                         # => [2..29]日
      # 29 日 23 時間 59 分 30 秒 <-> 59 日 23 時間 59 分 29 秒                   # => 約1ヶ月
      # 59 日 23 時間 59 分 30 秒 <-> 1 年 マイナス 31 秒                         # => [2..12]ヶ月
      # 1 年 マイナス 30 秒 <-> 2 年 マイナス 31 秒                               # => 約1年
      # 2 年 マイナス 30 秒 <-> time または date の最大値                         # => [2..X]年以上
      #
      # include_seconds = true で間隔 < 1 分 29 秒のとき
      # 0-4   秒      # => 5秒以内
      # 5-9   秒      # => 10秒以内
      # 10-19 秒      # => 20秒以内
      # 20-39 秒      # => 30秒前後
      # 40-59 秒      # => 1分以内
      # 60-89 秒      # => 1分
      #
      # Examples:
      #
      #   from_time = Time.now
      #   distance_of_time_in_words(from_time, from_time + 50.minutes) # => 約1時間
      #   distance_of_time_in_words(from_time, from_time + 15.seconds) # => 1分以内
      #   distance_of_time_in_words(from_time, from_time + 15.seconds, true) # => 20秒以内
      #
      # 注意： Rails は1年を365.25日として計算する。
      def distance_of_time_in_words(from_time, to_time = 0, include_seconds = false)
        from_time = from_time.to_time if from_time.respond_to?(:to_time)
        to_time = to_time.to_time if to_time.respond_to?(:to_time)
        distance_in_minutes = (((to_time - from_time).abs)/60).round
        distance_in_seconds = ((to_time - from_time).abs).round

        case distance_in_minutes
          when 0..1
            return (distance_in_minutes == 0) ? '1分以内' : '1分' unless include_seconds
            case distance_in_seconds
              when 0..4   then '5秒以内'
              when 5..9   then '10秒以内'
              when 10..19 then '20秒以内'
              when 20..39 then '30秒前後'
              when 40..59 then '1分以内'
              else             '1分'
            end

          when 2..44           then "#{distance_in_minutes}分"
          when 45..89          then '約1時間'
          when 90..1439        then "約#{(distance_in_minutes.to_f / 60.0).round}時間"
          when 1440..2879      then '1日'
          when 2880..43199     then "#{(distance_in_minutes / 1440).round}日"
          when 43200..86399    then '約1ヶ月'
          when 86400..525959   then "#{(distance_in_minutes / 43200).round}ヶ月"
          when 525960..1051919 then '約1年'
          else                      "#{(distance_in_minutes / 525960).round}年以上"
        end
      end

      def select_second_with_jp_time_unit(datetime, options = {}, html_options = {})
        select_second_without_jp_time_unit(datetime, options, html_options).chomp + (options[:use_jp_second] != false && !options[:use_hidden] ? '秒' : '') + "\n"
      end

      def select_minute_with_jp_time_unit(datetime, options = {}, html_options = {})
        select_minute_without_jp_time_unit(datetime, options, html_options).chomp + (options[:use_jp_minute] != false && !options[:use_hidden] ? '分' : '') + "\n"
      end

      def select_hour_with_jp_time_unit(datetime, options = {}, html_options = {})
        select_hour_without_jp_time_unit(datetime, options, html_options).chomp + (options[:use_jp_hour] != false && !options[:use_hidden] ? '時' : '') + "\n"
      end

      def select_day_with_jp_time_unit(date, options = {}, html_options = {})
        select_day_without_jp_time_unit(date, options, html_options).chomp + (options[:use_jp_day] != false && !options[:use_hidden] ? '日' : '') + "\n"
      end

      def select_month_with_jp_time_unit(date, options = {}, html_options = {})
        options.update(:use_month_numbers => true) if options[:use_jp_month] != false
        select_month_without_jp_time_unit(date, options, html_options).chomp  + (options[:use_jp_month] != false && !options[:use_hidden] ? '月' : '') + "\n"
      end

      def select_year_with_jp_time_unit(date, options = {}, html_options = {})
       # select_year_without_jp_time_unit(date, options, html_options).chomp  + (options[:use_jp_year] != false && !options[:use_hidden] ? '年' : '') + "\n"
        val = date ? (date.kind_of?(Fixnum) ? date : date.year) : ''
        if options[:use_hidden]
          hidden_html(options[:field_name] || 'year', val, options)
        else
          year_options = []
          y = date ? (date.kind_of?(Fixnum) ? (y = (date == 0) ? Date.today.year : date) : date.year) : Date.today.year

          start_year, end_year = (options[:start_year] || y-5), (options[:end_year] || y+5)
          step_val = start_year < end_year ? 1 : -1
          start_year.step(end_year, step_val) do |year|
            year_caption = build_year_caption(year, options)
            year_options << ((val == year) ?
              content_tag(:option, year_caption, :value => year, :selected => "selected") :
              content_tag(:option, year_caption, :value => year)
            )
            year_options << "\n"
          end
          
          era_format = options[:era_format] || :ja_long
          postfix = options[:use_jp_year] == false || (options[:use_era_name] == true &&  era_format ==  :ja_long ) ? '' : '年'

          select_html(options[:field_name] || 'year', year_options.join, options, html_options).chomp + postfix + "\n"
        end

      end
      
      def build_year_caption(year, options)
        if options[:use_era_name] == true 
           build_year_with_era_name(year, options)
        else
           year
        end
      end
      
      def build_year_with_era_name(year, options)
        case options[:era_format]
        when :ja_short
          era_formats = {:M => "明%d", :T => "大%d", :S => "昭%d", :H => "平%d"}
          era_first_years = {:M => '明1', :T => '明45/大1', :S => '大15/昭1', :H => '昭64/平1'}
        when :alphabet
          era_formats = {:M => "M%d", :T => "T%d", :S => "S%d", :H => "H%d"}
          era_first_years = {:M => 'M1', :T => 'M45/T1', :S => 'T15/S1', :H => 'S64/H1'}
        else # when :ja_long or others
          era_formats = {:M => "明治%d年", :T => "大正%d年", :S => "昭和%d年", :H => "平成%d年"}
          era_first_years = {:M => '明治元年', :T => '明45年/大1年', :S => '大15年/昭1年', :H => '昭64年/平1年'}
        end

        if year < 1868
          year
        elsif year == 1868
          era_first_years[:M]
        elsif year < 1912
          era_formats[:M] % (year - 1867)
        elsif year == 1912
          era_first_years[:T]
        elsif year < 1926
          era_formats[:T] % (year - 1911)
        elsif year == 1926
          era_first_years[:S]
        elsif year < 1989
          era_formats[:S] % (year - 1925)
        elsif year == 1989
          era_first_years[:H]
        else
          era_formats[:H] % (year - 1988)
        end
      end      

      alias_method_chain :select_minute, :jp_time_unit
      alias_method_chain :select_hour, :jp_time_unit
      alias_method_chain :select_second, :jp_time_unit
      alias_method_chain :select_day, :jp_time_unit
      alias_method_chain :select_month, :jp_time_unit
      alias_method_chain :select_year, :jp_time_unit
    end

    class InstanceTag
      def date_or_time_select_with_jp_time_unit(options, html_options = {})

        defaults = { :discard_type => true }
        options  = defaults.merge(options)
        datetime = value(object)
        datetime ||= default_time_from_options(options[:default]) unless options[:include_blank]

        position = { :year => 1, :month => 2, :day => 3, :hour => 4, :minute => 5, :second => 6 }

        order = (options[:order] ||= [:year, :month, :day])

        # Discard explicit and implicit by not being included in the :order
        discard = {}
        discard[:year]   = true if options[:discard_year] or !order.include?(:year)
        discard[:month]  = true if options[:discard_month] or !order.include?(:month)
        discard[:day]    = true if options[:discard_day] or discard[:month] or !order.include?(:day)
        discard[:hour]   = true if options[:discard_hour]
        discard[:minute] = true if options[:discard_minute] or discard[:hour]
        discard[:second] = true unless options[:include_seconds] && !discard[:minute]

        if datetime && discard[:day] && !discard[:month]
          datetime = datetime.change(:day => 1)
        end

        # Maintain valid dates by including hidden fields for discarded elements
        [:day, :month, :year].each { |o| order.unshift(o) unless order.include?(o) }
        # Ensure proper ordering of :hour, :minute and :second
        [:hour, :minute, :second].each { |o| order.delete(o); order.push(o) }

        date_or_time_select = ''
        order.reverse.each do |param|
          # Send hidden fields for discarded elements once output has started
          # This ensures AR can reconstruct valid dates using ParseDate
          next if discard[param] && date_or_time_select.empty?

          date_or_time_select.insert(0, self.send("select_#{param}", datetime, options_with_prefix(position[param], options.merge(:use_hidden => discard[param]))))
          date_or_time_select.insert(0,
            case param
             # assumes that the date format comes in the order of year, month, day, hour, minute and second.
              when :hour then (discard[:year] && discard[:day] ? "" : (options[:use_jp_day] == false ? " &mdash; " : " "))
              when :minute then (options[:use_jp_hour] == false ? " : "  : "")
              when :second then options[:include_seconds] ? (options[:use_jp_minute] == false ? " : "  : "") : ""
              else ""
            end)

        end

        date_or_time_select
      end

      alias_method_chain :date_or_time_select, :jp_time_unit
    end
  end
end